src/Nethermind/Nethermind.Trie/Pruning/TreePath.cs

```diff
@@ -0,0 +1,249 @@
+            return;
+        }
+
+        Span<byte> pathSpan = Span;
+
+        if (Length % 2 == 1)
+        {
+            this[Length] = nibbles[0];
+            Length++;
+            nibbles = nibbles[1..];
+        }
+
+        int byteLength = nibbles.Length / 2;
+        int pathSpanStart = Length / 2;
+        for (int i = 0; i < byteLength; i++)
+        {
+            pathSpan[i + pathSpanStart] = Nibbles.ToByte(nibbles[i * 2], nibbles[i * 2 + 1]);
+            Length += 2;
+        }
+
+        if (nibbles.Length % 2 == 1)
+        {
+            this[Length] = nibbles[^1];
+            Length++;
+        }
+    }
+
+    internal void AppendMut(byte nib)
+    {
+        this[Length] = nib;
+        Length++;
+    }
+
+    public byte this[int childPathLength]
+    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        readonly get
+        {
+            if (childPathLength >= 64) throw new IndexOutOfRangeException();
+            if (childPathLength % 2 == 0)
+            {
+                return (byte)((Span[childPathLength / 2] & 0xf0) >> 4);
+            }
+            else
+            {
+                return (byte)(Span[childPathLength / 2] & 0x0f);
+            }
+        }
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        set
+        {
+            if (childPathLength >= 64) throw new IndexOutOfRangeException();
+            if (childPathLength % 2 == 0)
+            {
+                Span[childPathLength / 2] =
+                    (byte)((Span[childPathLength / 2] & 0x0f) |
+                           (value & 0x0f) << 4);
+            }
+            else
+            {
+                Span[childPathLength / 2] =
+                    (byte)((Span[childPathLength / 2] & 0xf0) |
+                           ((value & 0x0f)));
+            }
+        }
+    }
+
+    public readonly TreePath Truncate(int pathLength)
+    {
+        TreePath copy = this;
+        copy.TruncateMut(pathLength);
+        return copy;
+    }
+
+    public void TruncateMut(int pathLength)
+    {
+        if (pathLength > Length) throw new IndexOutOfRangeException("path length must be less than current length");
+        if (pathLength == Length) return;
+
+        if (Length % 2 == 1)
+        {
+            this[Length - 1] = 0;
+            Length--;
+        }
+
+        if (pathLength == Length) return;
+
+        int byteClearStart = (pathLength + 1) / 2;
+        Span[byteClearStart..(Length / 2)].Clear();
+
+        if (pathLength % 2 == 1)
+        {
+            this[pathLength] = 0;
+        }
+
+        Length = pathLength;
+    }
+
+    public override string ToString()
+    {
+        return $"{Length} {Path}";
+    }
+
+    public static bool operator ==(in TreePath left, in TreePath right)
+    {
+        return left.Equals(right);
+    }
+
+    public static bool operator !=(TreePath left, TreePath right)
+    {
+        return !(left == right);
+    }
+
+    public bool Equals(TreePath other)
```

Comment:
As is a mutable struct but commonly passed as `in` mark all the non mutating methods as `readonly` or C# will make a defensive copy of the struct before calling the methods (as `in` says it should be a readonly reference pass)

Reply:
Done

Reply:
mean this should also do

`public bool Equals` => `public readonly bool Equals` to tell the C# compiler that no state is altered so it doesn't need to make a copy of the readonly struct before calling it to prevent against it mutating the one the reference points at

Reply:
Readonly instance members https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/struct#readonly-instance-members

You have it on some methods already

Reply:
Done